### PR TITLE
Fixing race condition between end-operation and future completion.

### PR DESCRIPTION
**Pull request by sdedic**
@mbien mentioned test failures in #7655, I've experienced them too in #7665 ([action run](https://github.com/apache/netbeans/actions/runs/10598938114/attempts/1)) and #7697 ([action run](https://github.com/apache/netbeans/actions/runs/10595857319/attempts/1)).

I've "sort of" traced/guessed it to finalization of a reload: in the original code, the Future completes (which means it unblocks the test thread), then the reload formally ends (pending reload is removed). The test may be able to observe the still-running reload (nearing its end) during its `tearDown`. The order was chosen so that events are fired only after Future completion, so potential observers see the new state as completed/ready. 

The PR splits the "formal termination" of a pending reload from event delivery, so the sequence is now:
- end the reload, potentially prepare the next one in the queue
- complete the reload Future
- deliver events and proceed with the next reload
This sequence is actually better / more predictable for clients.

### CODE REVIEW

**Comment by sdedic** [comment]
Still does not work ;) so I'll play with it ... will ping after it is reviewable.

**Comment by sdedic** [comment]
@mbien seems better now. The 2nd failure was actually a bad test diagnostics ... but while fixing I made sure that the events created during a reload are delivered before next reload starts.

**Comment by sdedic** [comment]
@mbien ping

**Review by mbien — CHANGES_REQUESTED** [review]
I ran maven and gradle tests 20 times in sequence using a custom workflow on my clone and got only one failure in the gradle step
https://github.com/mbien/netbeans/actions/runs/10656862483 gradle runs
https://github.com/mbien/netbeans/actions/runs/10659082827 maven runs (still running but it looks good so far)

so the fix seems to work I believe. Some comments inline.


```
2024-09-01T20:57:07.1157732Z     [junit] Testcase: testGeneratedSources(org.netbeans.modules.gradle.java.classpath.GradleSourcesImplTest):	FAILED
2024-09-01T20:57:07.1159927Z     [junit] expected:<2> but was:<0>
2024-09-01T20:57:07.1160727Z     [junit] junit.framework.AssertionFailedError: expected:<2> but was:<0>
2024-09-01T20:57:07.1162431Z     [junit] 	at org.netbeans.modules.gradle.java.classpath.GradleSourcesImplTest.testGeneratedSources(GradleSourcesImplTest.java:55)
2024-09-01T20:57:07.1164429Z     [junit] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2024-09-01T20:57:07.1166016Z     [junit] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
2024-09-01T20:57:07.1167834Z     [junit] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2024-09-01T20:57:07.1169325Z     [junit] 	at org.netbeans.junit.NbTestCase.access$200(NbTestCase.java:83)
2024-09-01T20:57:07.1170432Z     [junit] 	at org.netbeans.junit.NbTestCase$2.doSomething(NbTestCase.java:488)
2024-09-01T20:57:07.1171555Z     [junit] 	at org.netbeans.junit.NbTestCase$1Guard.run(NbTestCase.java:409)
2024-09-01T20:57:07.1172530Z     [junit] 	at java.base/java.lang.Thread.run(Thread.java:840)
```

**Comment by mbien** [comment]
I pressed the wrong button, wanted to comment and not request changes.

**Comment by mbien** [comment]
I suppose we can link https://github.com/apache/netbeans/issues/7263 to this PR too?

File: ide/project.dependency/src/org/netbeans/modules/project/dependency/reload/ProjectReloadInternal.java [hunk 1]
```diff
@@ -843,32 +891,34 @@ private void endOperation(Project p, Reloader reload) {
...
+
             RequestProcessor floader = loader;
 
-            CompletableFuture<ProjectState> f = CompletableFuture.runAsync(() -> nextReloader.initRound(), loader).
-                    thenCompose((v) -> nextReloader.start(floader));
-            f.whenComplete((a, b) -> {
+            CompletableFuture<ProjectState> f = CompletableFuture.runAsync(() -> fNextReloader.initRound(), loader).
+                    thenCompose((v) -> fNextReloader.start(floader));
+            // run this cleanup in the dispatcher thread
+            f.whenCompleteAsync((a, b) -> {
```
[the remark this answers is not in the document]
Comment (sdedic) [reply ide/project.dependency/src/org/netbeans/modules/project/dependency/reload/ProjectReloadInternal.java:900]: Understood, but separating the 1st part into a temporary turned out to be much nicer when debugging ;)

**Comment by sdedic** [comment]
rebased / squashed.

File: ide/project.dependency/src/org/netbeans/modules/project/dependency/reload/ProjectReloadInternal.java [hunk 2]
```diff
@@ -743,32 +743,62 @@ public void runProjectAction(Project p, Runnable r) {
...
+        // from stateIdentity, so just check they did not obtain another one:
+        releases.removeIf(expired -> {
+            ProjectStateData d = expired.state.get();
+            if (d == null) {
+                return true;
+            }
+            IdentityHolder h = stateIdentity.get(d);
+            return h != null && h != expired;
+        });
+        op.releases = new ArrayList<>();
```
Comment (mbien) [inline ide/project.dependency/src/org/netbeans/modules/project/dependency/reload/ProjectReloadInternal.java:777]: moving L777 to L766 would make this a little bit safer.

**Review by mbien — APPROVED** [review]

File: ide/project.dependency/src/org/netbeans/modules/project/dependency/reload/ProjectReloadInternal.java (hunk 2 above)
Comment (sdedic) [reply ide/project.dependency/src/org/netbeans/modules/project/dependency/reload/ProjectReloadInternal.java:777]: IMHO not important as entire call executes under `synchronized(this)` which also guards other accesses to `op.releases`. Maybe optically safer = adjacent to taking out the value.

**Pull request merged**